### PR TITLE
feat(metrics): Implement basic histogram functions on metrics_v2 API + QueryBuilder [INGEST-947]

### PR DESCRIPTION
**Pull request by untitaker**
Expose:

* ~`raw_histogram`, which is just an alias for the clickhouse `histogram(250)` function~
* `histogram`, which is a "derived op" that gets a histogram from clickhouse, runs re-bucketing/trimming logic and returns the result
* `histogram`-related query parameters
* `includeSeries=True` to avoid creating a "series of histograms" (which would be expensive)
* `includeTotals=True` for consistency

The current plan is to expose histograms via discover first, such that we have an e2e prototype with UI asap. This, in contrast, exposes histograms entirely through the metrics API, while the consensus is to actually query histograms from a separate endpoint. I would propose that we still move forward with this such that we can try it out in production on real data. In a followup I'll hide those query parameters from the web API and build the saparate endpoint.

### CODE REVIEW

File: src/sentry/snuba/metrics/fields/base.py [hunk 1]
```diff
@@ -233,7 +366,8 @@ class DerivedMetricDefinition:
     unit: str
     op: Optional[str] = None
     result_type: Optional[MetricType] = None
-    snql: Optional[Function] = None
+    # TODO: better typing
+    snql: Optional[Callable[..., Function]] = None
```
Comment (untitaker) [inline src/sentry/snuba/metrics/fields/base.py:370]: I had to fix this because `snql` is a python function that returns a snql function expression. python functions are `Callable`, `Function` is from snuba-sdk

File: src/sentry/snuba/metrics/fields/base.py [hunk 2]
```diff
@@ -247,14 +381,19 @@ def _raise_entity_validation_exception(self, func_name: str):
...
 class SingularEntityDerivedMetric(DerivedMetric):
-    def __init__(self, *args: Any, **kwargs: Any) -> None:
-        super().__init__(*args, **kwargs)  # type: ignore
-        self.result_type = "numeric"
-
-        if self.snql is None:
-            raise DerivedMetricParseException(
-                "SnQL cannot be None for instances of SingularEntityDerivedMetric"
-            )
+    # Pretend for the typechecker that __init__ is not overridden, such that
```
Comment (untitaker) [inline src/sentry/snuba/metrics/fields/base.py:384]: With this patch the mis-typing of `snql` is exposed. Without it the IDE doesn't error, because `args/kwargs` is weakly typed.

File: src/sentry/snuba/metrics/fields/base.py [hunk 3]
```diff
@@ -364,13 +503,16 @@ def generate_default_null_values(self):
     def generate_available_operations(self):
         return []
 
-    def run_post_query_function(self, data, idx=None):
+    def run_post_query_function(self, data, query_definition: QueryDefinition, idx=None):
```
Comment (untitaker) [inline src/sentry/snuba/metrics/fields/base.py:506]: I have to add this parameter such that we can access histogram opts. since some fields are instantiated as singletons, there's no other way than to pass it as params

File: src/sentry/snuba/sessions_v2.py [hunk 4]
```diff
@@ -138,7 +138,7 @@ def extract_from_row(self, row, group):
 
 
 def finite_or_none(val):
-    if not math.isfinite(val):
+    if isinstance(val, (int, float)) and not math.isfinite(val):
```
Comment (untitaker) [inline src/sentry/snuba/sessions_v2.py:141]: the return value of histogram is an array of arrays. `math.isfinite` will crash on that value

File: src/sentry/snuba/metrics/utils.py [hunk 5]
```diff
@@ -54,7 +54,7 @@
 #: The type of metric, which determines the snuba entity to query
 MetricType = Literal["counter", "set", "distribution", "numeric"]
 
-MetricEntity = Literal["metrics_counters", "metrics_sets", "metrics_distribution"]
+MetricEntity = Literal["metrics_counters", "metrics_sets", "metrics_distributions"]
```
Comment (untitaker) [inline src/sentry/snuba/metrics/utils.py:57]: Fixed that as my ide was complaining

File: src/sentry/snuba/metrics/query_builder.py [hunk 6]
```diff
@@ -581,25 +591,41 @@ def translate_results(self):
...
         # required nor expected in the response
         for group in groups:
-            totals, series = group["totals"], group["series"]
+            totals = group["totals"]
+            if self._query_definition.include_series:
+                series = group["series"]
+            else:
+                series = None
+
+            # XXX(markus): This deepcopy should not be necessary
```
Comment (untitaker) [inline src/sentry/snuba/metrics/query_builder.py:628]: - [ ] I forgot, let me try to remove that one later

File: src/sentry/snuba/metrics/query_builder.py [hunk 7]
```diff
@@ -581,25 +591,41 @@ def translate_results(self):
 
         # Applying post query operations for totals and series
         for group in groups:
-            totals, series = group["totals"], group["series"]
+            totals = group["totals"]
+            if self._query_definition.include_series:
```
Comment (untitaker) [inline src/sentry/snuba/metrics/query_builder.py:595]: bunch of if-stmts scattered around to avoid computing series...

File: src/sentry/snuba/metrics/query_builder.py [hunk 8]
```diff
@@ -440,24 +450,21 @@ def _build_queries(self, query_definition):
...
                 where=where + where_for_entity,
                 groupby=groupby,
                 orderby=orderby,
-                limit=query_definition.limit,
-                offset=query_definition.offset,
-                rollup=query_definition.rollup,
-                intervals_len=len(list(get_intervals(query_definition))),
+                limit=self._query_definition.limit,
+                offset=self._query_definition.offset,
+                rollup=self._query_definition.rollup,
```
Comment (untitaker) [inline src/sentry/snuba/metrics/query_builder.py:463]: those params can now probably go away, since query_definition is on the instance

File: src/sentry/snuba/metrics/query_builder.py [hunk 9]
```diff
@@ -349,6 +353,10 @@ def _build_totals_and_series_queries(
             granularity=Granularity(rollup),
             orderby=orderby,
         )
+
+        if not self._query_definition.include_series:
```
Comment (untitaker) [inline src/sentry/snuba/metrics/query_builder.py:357]: I needed query_definition here. Since I saw that we're already passing around quite a lot of state between static methods, I thought it would be appropriate to refactor this class such that it stores the query definition in `__init__` and does everything else lazily

File: src/sentry/snuba/metrics/fields/base.py [hunk 10]
```diff
@@ -658,16 +772,8 @@ def generate_bottom_up_dependency_tree_for_metrics(query_definition_fields_set):
...
-        # We are only interested in the dependency tree from instances of
-        # CompositeEntityDerivedMetric as they don't have a direct mapping to SnQL and so
-        # need to be computed post query which is practically when this function is called
-        if isinstance(derived_metric, CompositeEntityDerivedMetric):
-            dependency_list.extend(derived_metric.generate_bottom_up_derived_metrics_dependencies())
-        elif isinstance(derived_metric, SingularEntityDerivedMetric):
-            dependency_list.append((None, derived_metric.metric_name))
+    for op, metric_name in query_definition_fields_set:
+        dependency_list.extend(
+            metric_object_factory(op, metric_name).generate_bottom_up_derived_metrics_dependencies()
```
Comment (untitaker) [inline src/sentry/snuba/metrics/fields/base.py:777]: I added this method to the FieldBase class so we don't have to do instance checks here. Also I had to make the lookup fo through metric_object_factory because otherwise the derived op was not detected here

File: src/sentry/snuba/metrics/fields/base.py [hunk 11]
```diff
@@ -642,10 +748,18 @@ class DerivedMetricKey(Enum):
...
+DERIVED_OPS = {
+    "histogram": lambda metric_name: HistogramMetricField(op="histogram", metric_name=metric_name),
+}
+
 
 def metric_object_factory(op: Optional[str], metric_name: str) -> MetricFieldBase:
     """Returns an appropriate instance of MetricsFieldBase object"""
-    if metric_name in DERIVED_METRICS:
+    if op in DERIVED_OPS and metric_name in DERIVED_METRICS:
+        raise InvalidParams("derived ops cannot be used on derived metrics")
```
Comment (untitaker) [inline src/sentry/snuba/metrics/fields/base.py:759]: I think this might be an important limitation in the future, not sure

File: src/sentry/snuba/metrics/fields/base.py [hunk 12]
```diff
@@ -431,45 +573,9 @@ def __recursively_generate_singular_entity_constituents(cls, projects, derived_m
...
-        [
-        "session.errored_set", "session.errored_preaggregated",
-        "composite_entity_derived_metric", "session.errored"
-        ]
-        And this order is necessary because we would loop over this list to compute
-        `composite_entity_derived_metric` first which does not have a direct mapping to a query
-        alias before we are able to compute `session.errored` which also does not have a direct
-        query alias
-        """
+        # We are only interested in the dependency tree from instances of
```
Comment (untitaker) [inline src/sentry/snuba/metrics/fields/base.py:576]: I moved some doc blocks around because this method now also lives in the baseclass

File: src/sentry/snuba/metrics/fields/base.py [hunk 13]
```diff
@@ -221,10 +284,80 @@ def generate_available_operations(self):
...
+                rv_height = overlap_perc * height
+                rv[lower2, upper2] += rv_height
+
+        return [(lower, upper, height) for (lower, upper), height in rv.items()]
+
+    def run_post_query_function(self, data, query_definition: QueryDefinition, idx=None):
+        key = f"{self.op}({self.metric_name})"
+        if idx is None:
+            data[key] = {
+                "histogram_metric_test": self.__resize_histogram(data[key], query_definition)
```
Comment (untitaker) [inline src/sentry/snuba/metrics/fields/base.py:353]: - [ ] whoops I need to remove that wrapped dictionary

File: src/sentry/snuba/metrics/fields/base.py [hunk 14]
```diff
@@ -221,10 +284,80 @@ def generate_available_operations(self):
...
+            (
+                min_val + (max_val - min_val) * (x / bucket_count),
+                min_val + (max_val - min_val) * ((x + 1) / bucket_count),
+            )
+            for x in range(bucket_count)
+        ]
+
+        rv = {bucket: 0 for bucket in buckets}
+
+        # XXX: quadratic function
```
Comment (untitaker) [inline src/sentry/snuba/metrics/fields/base.py:334]: this hideous function was copied from the histogram experiments repo. I believe it will perform fine for now because we're going to do `250*250=62k` iterations at max, which is probably still manageable in endpoints

File: src/sentry/snuba/metrics/fields/base.py [hunk 15]
```diff
@@ -55,6 +69,12 @@
 )
 from sentry.utils.snuba import raw_snql_query
 
+if TYPE_CHECKING:
+    from sentry.snuba.metrics.query_builder import QueryDefinition
```
Comment (untitaker) [inline src/sentry/snuba/metrics/fields/base.py:73]: I have to do this and the `__future__` import because of circular imports

File: tests/sentry/snuba/metrics/test_query_builder.py [hunk 16]
```diff
@@ -796,3 +796,23 @@ def test_translate_results_missing_slots(_1, _2, monkeypatch):
             },
         },
     ]
+
+
+def test_histogram_snuba_query(monkeypatch):
```
Comment (untitaker) [inline tests/sentry/snuba/metrics/test_query_builder.py:801]: - [ ] this should not have made it in

**Review by ahmedetefy — COMMENTED** [review]
I think after our call, this looks good to me except for the minor comments we discussed.

The only minor thing I would probably do is refactor the histogram post query function to its module/function so its isolated and easier to test

Other than that, all good

File: src/sentry/snuba/metrics/fields/base.py [hunk 17]
```diff
@@ -221,10 +284,76 @@ def generate_available_operations(self):
...
+    ) -> List[OrderBy]:
+        raise DerivedMetricParseException("histograms cannot be ordered")
+
+    def generate_default_null_values(self):
+        return []
+
+    def generate_available_operations(self):
+        return [self.op]
+
+    def __resize_histogram(
```
Comment (jjbayer) [inline src/sentry/snuba/metrics/fields/base.py:307]: nit: Would `__rebucket_histogram` be a better name?

File: src/sentry/snuba/metrics/fields/base.py (hunk 14 above)
Comment (jjbayer) [reply src/sentry/snuba/metrics/fields/base.py:334]: Should we create a span around this loop so we can measure how long it takes?

File: src/sentry/snuba/metrics/fields/base.py [hunk 18]
```diff
@@ -221,10 +284,76 @@ def generate_available_operations(self):
...
+            max_val = min(max_val, query_definition.histogram_to)
+        bucket_count = query_definition.histogram_buckets
+
+        buckets = [
+            (
+                min_val + (max_val - min_val) * (x / bucket_count),
+                min_val + (max_val - min_val) * ((x + 1) / bucket_count),
+            )
+            for x in range(bucket_count)
+        ]
```
Comment (jjbayer) [inline src/sentry/snuba/metrics/fields/base.py:330]: nit: readability

```suggestion
       target_bucket_width = (max_val - min_val) / bucket_count
        buckets = [
            (
                min_val + i * target_bucket_width,
                min_val + (i + 1) * target_bucket_width,
            )
            for i in range(bucket_count)
        ]
```

File: src/sentry/snuba/metrics/fields/base.py [hunk 19]
```diff
@@ -221,10 +284,76 @@ def generate_available_operations(self):
...
+
+        rv = {bucket: 0 for bucket in buckets}
+
+        # XXX: quadratic function
+        assert len(data) < 300
+        assert len(rv) < 300
+        for lower, upper, height in data:
+            for lower2, upper2 in rv:
```
Comment (jjbayer) [inline src/sentry/snuba/metrics/fields/base.py:338]: nit: rename to `lower_target` / `lower_dest` / something similar?

File: src/sentry/snuba/metrics/fields/base.py (hunk 3 above)
Comment (jjbayer) [reply src/sentry/snuba/metrics/fields/base.py:502]: I feel like the histogram parameters should be passed to the `op` itself, i.e. `histogram(metric_name, from, to, buckets)`. Then you would be able to query multiple histograms with different parameters in one request, e.g. `?field=histogram(slow_thing, 0, 10000, 250)&field=histogram(fast_thing, 0, 100, 250)`.

**Review by jjbayer — APPROVED** [review]

File: src/sentry/snuba/metrics/fields/base.py (hunk 3 above)
Comment (untitaker) [reply src/sentry/snuba/metrics/fields/base.py:502]: we had to decide against that as it would mean writing a proper parser for fields. I also don't think this will be hard to support in the future

File: src/sentry/snuba/metrics/fields/base.py [hunk 20]
```diff
@@ -55,6 +60,21 @@
 )
 from sentry.utils.snuba import raw_snql_query
 
+__all__ = (
```
Comment (ahmedetefy) [inline src/sentry/snuba/metrics/fields/base.py:63]: Out of curiosity, why did you move this down?

File: src/sentry/snuba/metrics/fields/base.py [hunk 21]
```diff
@@ -219,12 +282,43 @@ def generate_available_operations(self):
...
+        return [(self.op, self.metric_name)]
+
+
+class HistogramMetricField(RawAggregatedMetric):
+    def generate_orderby_clause(
+        self, projects: Sequence[Project], direction: Direction
+    ) -> List[OrderBy]:
+        raise DerivedMetricParseException("histograms cannot be ordered")
+
+    def run_post_query_function(self, data, query_definition: QueryDefinition, idx=None):
```
Comment (ahmedetefy) [inline src/sentry/snuba/metrics/fields/base.py:301]: This is neat! Thanks

File: src/sentry/snuba/metrics/query_builder.py [hunk 22]
```diff
@@ -566,14 +578,11 @@ def _extract_data(self, data, groups):
...
-            for data in totals:
-                self._extract_data(data, groups)
-
-            if "series" in subresults:
-                series = subresults["series"]["data"]
-                for data in series:
-                    self._extract_data(data, groups)
+            for k in "totals", "series":
+                if k in subresults:
+                    totals = subresults[k]["data"]
```
Comment (ahmedetefy) [inline src/sentry/snuba/metrics/query_builder.py:583]: nit: I would name this something other than `totals` cause it is a bit confusing

File: src/sentry/snuba/metrics/query_builder.py [hunk 23]
```diff
@@ -590,26 +599,36 @@ def translate_results(self):
...
         # from the generated dependency tree. These metrics that are to be removed were added to
         # be able to generate fields that require further processing post query, but are not
         # required nor expected in the response
         for group in groups:
-            totals, series = group["totals"], group["series"]
-            for key in copy.deepcopy(list(totals.keys())):
+            totals = group.get("totals")
+            series = group.get("series")
+
+            for key in set(totals or ()) | set(series or ()):
```
Comment (ahmedetefy) [inline src/sentry/snuba/metrics/query_builder.py:631]: I don't understand this loop condition.. What does this do?

File: src/sentry/snuba/metrics/fields/histogram.py [hunk 24]
```diff
@@ -0,0 +1,58 @@
+from typing import List, Optional, Tuple
+
+import sentry_sdk
+
+ClickhouseHistogram = List[Tuple[float, float, float]]
+
+
+def rebucket_histogram(
```
Comment (ahmedetefy) [inline src/sentry/snuba/metrics/fields/histogram.py:8]: I think we should probably add tests for this module

File: src/sentry/snuba/metrics/fields/base.py (hunk 20 above)
Comment (untitaker) [reply src/sentry/snuba/metrics/fields/base.py:63]: I didn't, pre-commit did

File: src/sentry/snuba/metrics/query_builder.py [hunk 25]
```diff
@@ -179,9 +175,12 @@ def __init__(self, query_params, paginator_kwargs: Optional[Dict] = None):
             raise InvalidField(
                 "We don't have more than 250 buckets stored for any given metric bucket."
             )
-        self.histogram_from = query_params.get("histogramFrom", None)
-        self.histogram_to = query_params.get("histogramTo", None)
+        histogram_from = query_params.get("histogramFrom", None)
+        histogram_to = query_params.get("histogramTo", None)
+        self.histogram_from = int(histogram_from) if histogram_from is not None else None
+        self.histogram_to = int(histogram_to) if histogram_to is not None else None
```
Comment (jjbayer) [inline src/sentry/snuba/metrics/query_builder.py:181]: `rebucket_histograms` take floats, so I guess these should be floats, too? Or is there a special reason to only accept ints?

```suggestion
        self.histogram_from = float(histogram_from) if histogram_from is not None else None
        self.histogram_to = float(histogram_to) if histogram_to is not None else None
```

Are we fine with responding with status code 500 when I pass `histogramFrom=foo`?

File: src/sentry/snuba/metrics/query_builder.py (hunk 23 above)
Comment (untitaker) [reply src/sentry/snuba/metrics/query_builder.py:637]: I thought I responded to this... basically since `totals` can be `None`, we can't use it to iterate over existing fields. this takes both dicts, and unites their keys into a single set. `set(None)` is invalid so I use `set(totals or ())` to create an empty set if `totals` is `None`.

File: src/sentry/snuba/metrics/query_builder.py (hunk 22 above)
Comment (untitaker) [reply src/sentry/snuba/metrics/query_builder.py:589]: done

File: src/sentry/snuba/metrics/query_builder.py (hunk 25 above)
Comment (untitaker) [reply src/sentry/snuba/metrics/query_builder.py]: >Are we fine with responding with status code 500 when I pass histogramFrom=foo?

I think we should fix those cases once ppl complain.. there are too many

**Review by ahmedetefy — APPROVED** [review]
lgtm! Thanks for changes
